### PR TITLE
Page tables

### LINKED ISSUE #445: Consider using a page table for granular memory protection

### PR DESCRIPTION
**Pull request by Gaberboo**
Resolves #445
Maps executable sections with a page table

### CODE REVIEW

File: payload/sp/security/PageTable.c [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        Dol,
+        Rel,
+        PayloadText,
+        PayloadReplacements,
+        0
+    };
+
+    u32 HTABORG = (u32)pageTable & HTABORG_MASK;
+    SetSDR1(VIRTUAL_TO_PHYSICAL(HTABORG) | HTABMASK);
+        for (size_t index = 0;pageTableMaps[index];index++) {
```
Comment (riidefi) [inline payload/sp/security/PageTable.c:125]: Run this file through clang-format again

File: payload/sp/security/PageTable.c [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+    u32 primaryHash = VSID ^ pageIndex;
+    u32 secondaryHash = ~primaryHash;
+    u32 hashArray[] = {
+            primaryHash,
+            secondaryHash,
+    };
+    static_assert((sizeof(hashArray) / sizeof(u32)) == 2);
+
+    for (size_t hashIndex = 0; hashIndex < (sizeof(hashArray) / sizeof(u32)); hashIndex++) {
```
Comment (riidefi) [inline payload/sp/security/PageTable.c:146]:
```diff
- (sizeof(hashArray) / sizeof(u32))
+ ARRAY_SIZE(hashArray)
```

File: payload/sp/security/PageTable.c [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        u32 index = ((hashArray[hashIndex] & (HTABMASK << 10)) | (hashArray[hashIndex] & 0x3FF)) << 6;
+        PageTableEntry *pteg = (PageTableEntry *)(pageTable + index);
+
+        for (int entryIndex = 0; entryIndex < 8; entryIndex++) {
+            PageTableEntry *pte = pteg + entryIndex;
+            if (pte->V) {
+                continue;
+            }
```
Comment (riidefi) [inline payload/sp/security/PageTable.c:154]: I'd pull out this linear search and unindent the block below:
```c
PageTableEntry *pte = nullptr;
for (size_t entryIndex = 0; entryIndex < 8; entryIndex++) {
    PageTableEntry *it= &pteg[entryIndex];
    if (!it->V) { pte = it; break; }
}
if (!pte) { continue; }
```
If we move this file to C++ we could write this as
```cpp
auto pte = std::find_if(pteg, pteg + 8, [](auto &x){ return !x.V; });
if (pte == pteg + 8) { continue; }
```

File: payload/sp/security/PageTable.c [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    u32 primaryHash = VSID ^ pageIndex;
+    u32 secondaryHash = ~primaryHash;
+    u32 hashArray[] = {
+            primaryHash,
+            secondaryHash,
+    };
+    static_assert((sizeof(hashArray) / sizeof(u32)) == 2);
+
+    for (size_t hashIndex = 0; hashIndex < (sizeof(hashArray) / sizeof(u32)); hashIndex++) {
+        u32 index = ((hashArray[hashIndex] & (HTABMASK << 10)) | (hashArray[hashIndex] & 0x3FF)) << 6;
```
Comment (riidefi) [inline payload/sp/security/PageTable.c:147]: `offset` is probably a better name here given sizeof(PageTableEntry) > 1

File: payload/sp/security/PageTable.c (hunk 1 above)
Comment (MikeIsAStar) [reply payload/sp/security/PageTable.c]: Resolved in commit [6e880f1](https://github.com/stblr/mkw-sp/pull/495/commits/6e880f1f8d09e43cccc045944ed8f9eaf5fee7db).

File: payload/sp/security/PageTable.c (hunk 2 above)
Comment (MikeIsAStar) [reply payload/sp/security/PageTable.c]: Resolved in commit [6e880f1](https://github.com/stblr/mkw-sp/pull/495/commits/6e880f1f8d09e43cccc045944ed8f9eaf5fee7db).

File: payload/sp/security/PageTable.c (hunk 3 above)
Comment (MikeIsAStar) [reply payload/sp/security/PageTable.c]: Resolved in commit [6e880f1](https://github.com/stblr/mkw-sp/pull/495/commits/6e880f1f8d09e43cccc045944ed8f9eaf5fee7db).

File: payload/sp/security/PageTable.c (hunk 4 above)
Comment (MikeIsAStar) [reply payload/sp/security/PageTable.c]: Resolved in commit [6e880f1](https://github.com/stblr/mkw-sp/pull/495/commits/6e880f1f8d09e43cccc045944ed8f9eaf5fee7db).

File: payload/sp/security/PageTable.cc [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                .WIMG = wimg,
+                .PP = pp,
+        };
+
+        pte->hilo[1] = entry.hilo[1];
+        asm volatile("eieio");
+        pte->hilo[0] = entry.hilo[0];
+        asm volatile("sync");
+
+        return true;
```
Comment (riidefi) [inline payload/sp/security/PageTable.cc:137]: Maybe pull out of loop

File: payload/sp/security/PageTable.cc [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    InitPageTable();
+
+    for (const PageTableEntryInfo &pteInfo : pteInfoArray) {
+        char *physicalStartAddress = (char *)ROUND_DOWN(pteInfo.physicalAddress, PAGE_SIZE);
+        char *effectiveStartAddress = (char *)ROUND_DOWN(pteInfo.effectiveAddress, PAGE_SIZE);
+        char *effectiveEndAddress =
+                (char *)ROUND_UP((char *)pteInfo.effectiveAddress + pteInfo.size, PAGE_SIZE);
+        u32 size = effectiveEndAddress - effectiveStartAddress;
+
+        for (u32 offset = 0; offset != size; offset += PAGE_SIZE) {
```
Comment (riidefi) [inline payload/sp/security/PageTable.cc:181]: Maybe assert PAGE_SIZE divides size

File: payload/sp/security/Memory.c [hunk 7]
```diff
@@ -2,6 +2,21 @@
 
 #include <revolution/os.h>
 
+#define MemoryRoundDown1024B(address) (((u32)(address)) & ~0x3FF)
+#define MemoryRoundUp1024B(address) (((u32)(address) + 0x3FF) & ~0x3FF)
+
+extern void InvalidateAllIBATs(void);
```
Comment (stblr) [inline payload/sp/security/Memory.c:8]: No need for `extern` here.

File: payload/sp/security/PageTable.cc [hunk 8]
```diff
@@ -0,0 +1,203 @@
+extern "C" {
+#include "PageTable.h"
+}
+
+#include <Common.hh>
+#include <algorithm>
```
Comment (stblr) [inline payload/sp/security/PageTable.cc:6]: https://github.com/stblr/mkw-sp/issues/245#issuecomment-1213490665

File: payload/sp/security/PageTable.cc [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+#define PAGE_TABLE_MEMORY_POWER_OF_2 16 // 64 Kibibytes
+static_assert(PAGE_TABLE_MEMORY_POWER_OF_2 > 15 && PAGE_TABLE_MEMORY_POWER_OF_2 < 26);
+#define PAGE_TABLE_SIZE (1 << PAGE_TABLE_MEMORY_POWER_OF_2)
+#define PAGE_SIZE (1 << 12)
+
+#define HTABMASK ~(~0 << (PAGE_TABLE_MEMORY_POWER_OF_2 - 16))
+#define HTABORG_MASK 0xFFFF0000
+
+typedef enum WIMG {
```
Comment (stblr) [inline payload/sp/security/PageTable.cc:27]: No need for `typedef` here and for the other definitions in C++.

File: payload/sp/security/PageTable.cc [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        u32 WIMG :  4;
+        u32      :  1;
+        u32 PP   :  2;
+        // clang-format on
+    };
+
+    u32 hilo[2];
+} PageTableEntry;
+
+__attribute__((aligned(PAGE_TABLE_SIZE))) static char pageTable[PAGE_TABLE_SIZE];
```
Comment (stblr) [inline payload/sp/security/PageTable.cc:76]: `alignas`

**Review by stblr — CHANGES_REQUESTED** [review]

File: payload/sp/security/PageTable.cc [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+            pte = it;
+            break;
+        }
+    }
+
+    if (!pte) {
+        return false;
+    }
+
+    const PageTableEntry entry = {
```
Comment (riidefi) [inline payload/sp/security/PageTable.cc:128]: Question: Are the reserved bits guaranteed to be zero here?

Comment (Gaberboo) [reply payload/sp/security/PageTable.cc:128]: According to Ghidra, yes
`          entry[1] = (RPN & 0xfffff000) + offset | (*(uint *)((int)puVar6 + 0xc) & 0xf) << 3 |
                     *(uint *)((int)puVar6 + 0x10) & 3;`
offset is only ever set to multiples of 0x1000

**Review by stblr — APPROVED** [review]

File: payload/sp/security/PageTable.cc (hunk 11 above)
Comment (riidefi) [reply payload/sp/security/PageTable.cc:129]: My question was in a moral--not observational--sense. But it does seem they are guaranteed to be zero by the standard:
> If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate shall be initialized implicitly the same as objects that have static storage duration.

**Pull request merged**